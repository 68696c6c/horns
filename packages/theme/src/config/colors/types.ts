export enum Mode {
  Light = 'light',
  Dark = 'dark',
}

export enum Color {
  Primary = 'primary',
  Secondary = 'secondary',
  Tertiary = 'tertiary',

  Action = 'action',
  Prominent = 'prominent',
  Selected = 'selected',

  Success = 'success',
  Info = 'info',
  Warning = 'warning',
  Danger = 'danger',

  Dark = 'dark',
  Neutral = 'neutral',
  Light = 'light',

  BgPrimary = 'bgPrimary',
  BgSecondary = 'bgSecondary',
  BgInverse = 'bgInverse',
}

export type BaseBrandColor = Color.Primary | Color.Secondary | Color.Tertiary

export type NotificationColor =
  | Color.Success
  | Color.Info
  | Color.Warning
  | Color.Danger

export type BaseColor =
  | BaseBrandColor
  | NotificationColor
  | Color.Dark
  | Color.Neutral
  | Color.Light

export type BrandColor =
  | Color.Primary
  | Color.Secondary
  | Color.Tertiary
  | Color.Action
  | Color.Prominent
  | Color.Selected

export type SurfaceColor = Color.BgPrimary | Color.BgSecondary

export enum Swatch {
  Base = 'base',
  Alt = 'alt',
  Readable = 'readable',
}

export type Swatches = {
  [key in Swatch]: string
}
