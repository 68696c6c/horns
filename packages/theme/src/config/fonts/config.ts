import { FontStylesConfig } from './fonts'
import {
  defaultDecorations,
  Decorations,
  Direction,
  defaultFontFamilies,
  FontFamilies,
  defaultFontSizes,
  FontSizes,
  defaultFontWeights,
  FontWeights,
  defaultLettings,
  Lettings,
  defaultTextSpacings,
  TextSpacings,
  Decoration,
  FontWeight,
  TextSpacing,
  FontFamily,
  TextAlign,
  Letting,
  FontStyleCss,
  BaseFontSize,
} from './types'

export interface Config {
  direction: Direction
  families: FontFamilies
  decorations: Decorations
  weights: FontWeights
  sizes: FontSizes
  spacing: TextSpacings
  letting: Lettings
  styles: FontStylesConfig
}

export const defaultConfig: Config = {
  direction: Direction.LTR,
  families: defaultFontFamilies,
  decorations: defaultDecorations,
  weights: defaultFontWeights,
  sizes: defaultFontSizes,
  spacing: defaultTextSpacings,
  letting: defaultLettings,
  styles: {
    text: {},
    compact: {
      letting: Letting.Min,
    },
    heading: {
      weight: FontWeight.Bold,
      spacing: TextSpacing.Max,
    },
    paragraph: {
      align: TextAlign.Justify,
      spacing: TextSpacing.Max,
    },
    control: {
      letting: Letting.Min,
    },
    label: {
      weight: FontWeight.Bold,
      letting: Letting.Min,
    },
    legal: {
      family: FontFamily.Secondary,
      style: FontStyleCss.Italic,
      size: BaseFontSize.Small,
    },
    link: {
      decoration: Decoration.Underline,
    },
    emphasized: {
      style: FontStyleCss.Italic,
    },
    strong: {
      weight: FontWeight.Bold,
    },
    quote: {
      style: FontStyleCss.Italic,
    },
    code: {
      family: FontFamily.Tertiary,
    },
  },
}
