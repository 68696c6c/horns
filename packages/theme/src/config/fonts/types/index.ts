export { defaultDecorations, Decorations, Decoration } from './decorations'
export {
  defaultFontFamilies,
  FontFamilies,
  FontFamily,
  FontFamilyStyle,
  FontFamilyFallback,
} from './families'
export { defaultLettings, Lettings, Letting } from './lettings'
export {
  defaultFontSizes,
  FontSizes,
  FontSize,
  HeadingLevel,
  BaseFontSize,
} from './sizes'
export { defaultTextSpacings, TextSpacings, TextSpacing } from './spacings'
export {
  Direction,
  FontKerning,
  FontStyle as FontStyleCss,
  TextAlign,
  TextTransform,
} from './types'
export { defaultFontWeights, FontWeights, FontWeight } from './weights'
