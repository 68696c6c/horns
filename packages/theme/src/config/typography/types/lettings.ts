export enum Letting {
  Min = 'min',
  Base = 'base',
  Max = 'max',
}

export type Lettings = {
  [key in Letting]: string
}

export const defaultLettings: Lettings = {
  min: '1em',
  base: '1.6em',
  max: '2em',
}
