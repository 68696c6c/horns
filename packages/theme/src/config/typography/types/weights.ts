export enum FontWeight {
  Lighter = 'lighter',
  Light = 'light',
  Base = 'base',
  SemiBold = 'semiBold',
  Bold = 'bold',
  Bolder = 'bolder',
}

export type FontWeights = {
  [key in FontWeight]: number
}

export const defaultFontWeights: FontWeights = {
  base: 400,
  lighter: 100,
  light: 200,
  semiBold: 600,
  bold: 700,
  bolder: 900,
}
