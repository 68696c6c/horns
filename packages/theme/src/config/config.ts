import { DeepPartial, mergeConfig } from '../utils'

import * as Breakpoints from './breakpoints'
import * as Colors from './colors'
import * as Fonts from './fonts'
import * as Shadows from './shadows'
import * as Sizes from './sizes'

export interface Config {
  breakpoints: Breakpoints.Config
  colors: Colors.Config
  fonts: Fonts.Config
  shadows: Shadows.Config
  sizes: Sizes.Config
}

export interface Theme {
  readonly breakpoints: Breakpoints.Theme
  readonly colors: Colors.Theme
  readonly fonts: Fonts.Theme
  readonly shadows: Shadows.Theme
  readonly sizes: Sizes.Theme
}

export const defaultConfig = {
  breakpoints: Breakpoints.defaultConfig,
  colors: Colors.defaultConfig,
  fonts: Fonts.defaultConfig,
  shadows: Shadows.defaultConfig,
  sizes: Sizes.defaultConfig,
}

export const make = (input?: DeepPartial<Config>): Theme => {
  const config = mergeConfig<Config>(defaultConfig, input)
  return {
    breakpoints: Breakpoints.make(config.breakpoints),
    colors: Colors.make(config.colors),
    fonts: Fonts.make(config.fonts),
    shadows: Shadows.make(config.shadows),
    sizes: Sizes.make(config.sizes),
  }
}
