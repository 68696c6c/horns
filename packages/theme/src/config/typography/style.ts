import { mergeConfig } from '../../utils'

import { Config } from './config'
import { FontConfig } from './fonts'
import { FontFamilyStyle, FontSize, HeadingLevel } from './types'

const computeFontFamily = (style: FontFamilyStyle): string =>
  [style.base, style.fallback].join(', ')

export interface FontStyle {
  family: string // font-family
  style: string // font-style
  weight: number // font-weight
  size: string // font-size
  align: string // text-align
  transform: string // text-transform
  decoration: {
    line: string
    style: string
  } // text-decoration
  kerning: string // font-kerning
  spacing: string // margin; the space before and after typographic block-level elements
  indent: string // text-indent
  letting: string // line-height - overrides font-family value
  tracking: string // word-spacing - overrides font-family value
}

const makeFontStyle = (
  font: FontConfig,
  config: Config,
  size?: FontSize,
): FontStyle => {
  const { families, decorations, weights, spacing, sizes, letting } = config
  const fontDecoration = decorations[font.decoration]
  return {
    family: computeFontFamily(families[font.family]),
    style: font.style,
    weight: weights[font.weight],
    size: sizes[size || font.size],
    align: font.align,
    transform: font.transform,
    decoration: {
      line: fontDecoration.line,
      style: fontDecoration.style || '',
    },
    kerning: font.kerning,
    spacing: spacing[font.spacing],
    indent: font.indent,
    letting: letting[font.letting],
    tracking: font.tracking,
  }
}

export interface FontStateArgs {
  fontStyle: Partial<FontConfig>
  fontBase: Readonly<FontConfig>
  config: Readonly<Config>
  size?: HeadingLevel
}

export const evalFontStyle = ({
  fontStyle,
  fontBase,
  config,
  size,
}: FontStateArgs): FontStyle => {
  const base = mergeConfig<FontConfig>(fontBase, fontStyle)
  return makeFontStyle(base, config, size)
}
