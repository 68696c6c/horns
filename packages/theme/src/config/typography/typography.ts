import { mergeConfig } from '../../utils'

import { Config, defaultConfig } from './config'
import { Font, FontConfig } from './fonts'
import { FontStyle, evalFontStyle } from './style'
import {
  BaseFontSize,
  Decoration,
  FontFamily,
  FontKerning,
  FontStyleCss,
  FontWeight,
  HeadingLevel,
  Letting,
  TextAlign,
  TextSpacing,
  TextTransform,
} from './types'

export const makeDefaultFontConfig = (config: Config): FontConfig => ({
  family: FontFamily.Primary,
  style: FontStyleCss.Normal,
  weight: FontWeight.Base,
  size: BaseFontSize.Base,
  align: config.direction === 'rtl' ? TextAlign.Right : TextAlign.Left,
  transform: TextTransform.None,
  decoration: Decoration.Base,
  kerning: FontKerning.Normal,
  spacing: TextSpacing.Base,
  indent: '',
  letting: Letting.Base,
  tracking: '',
})

export type Typography = Required<
  {
    [key in Font | HeadingLevel]: Readonly<FontStyle>
  }
>

export const makeTypography = (input?: Partial<Config>): Typography => {
  const config = mergeConfig<Config>(defaultConfig, input)
  const fontBase = makeDefaultFontConfig(config)

  const result: Partial<Typography> = {}
  Object.values(Font).forEach((font) => {
    const fontStyle = config.styles[font]
    result[font] = evalFontStyle({
      fontStyle,
      fontBase,
      config,
    })
  })
  const fontStyle = config.styles.heading
  result.h1 = evalFontStyle({
    fontStyle,
    fontBase,
    config,
    size: HeadingLevel.H1,
  })
  result.h2 = evalFontStyle({
    fontStyle,
    fontBase,
    config,
    size: HeadingLevel.H2,
  })
  result.h3 = evalFontStyle({
    fontStyle,
    fontBase,
    config,
    size: HeadingLevel.H3,
  })
  result.h4 = evalFontStyle({
    fontStyle,
    fontBase,
    config,
    size: HeadingLevel.H4,
  })
  result.h5 = evalFontStyle({
    fontStyle,
    fontBase,
    config,
    size: HeadingLevel.H5,
  })
  result.h6 = evalFontStyle({
    fontStyle,
    fontBase,
    config,
    size: HeadingLevel.H6,
  })

  return result as Typography
}
