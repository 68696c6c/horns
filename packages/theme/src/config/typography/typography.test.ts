import { defaultConfig } from './config'
import { makeDefaultFontConfig, makeTypography } from './typography'
import { Direction, TextAlign } from './types'

describe('makeTypography', () => {
  it('should match snapshot', () => {
    const result = makeTypography()
    expect(result).toMatchSnapshot()
  })
})

describe('makeDefaultFontConfig', () => {
  it('should support right-to-left writing direction', () => {
    const config = defaultConfig
    config.direction = Direction.RTL
    const result = makeDefaultFontConfig(config)
    expect(result.align).toEqual(TextAlign.Right)
  })
})
