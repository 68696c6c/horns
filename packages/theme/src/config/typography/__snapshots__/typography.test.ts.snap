// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeTypography should match snapshot 1`] = `
Object {
  "code": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Monaco, monospace",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "medium",
    "spacing": "",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 400,
  },
  "compact": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1em",
    "size": "medium",
    "spacing": "",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 400,
  },
  "control": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1em",
    "size": "medium",
    "spacing": "",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 400,
  },
  "emphasized": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "medium",
    "spacing": "",
    "style": "italic",
    "tracking": "",
    "transform": "none",
    "weight": 400,
  },
  "h1": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "3.21em",
    "spacing": "0 0 1em 0",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 700,
  },
  "h2": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "2.36em",
    "spacing": "0 0 1em 0",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 700,
  },
  "h3": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "1.64em",
    "spacing": "0 0 1em 0",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 700,
  },
  "h4": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "1.29em",
    "spacing": "0 0 1em 0",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 700,
  },
  "h5": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "1.15em",
    "spacing": "0 0 1em 0",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 700,
  },
  "h6": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "1em",
    "spacing": "0 0 1em 0",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 700,
  },
  "heading": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "medium",
    "spacing": "0 0 1em 0",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 700,
  },
  "label": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1em",
    "size": "medium",
    "spacing": "",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 700,
  },
  "legal": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Times New Roman, serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "0.75em",
    "spacing": "",
    "style": "italic",
    "tracking": "",
    "transform": "none",
    "weight": 400,
  },
  "link": Object {
    "align": "left",
    "decoration": Object {
      "line": "underline",
      "style": "solid",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "medium",
    "spacing": "",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 400,
  },
  "paragraph": Object {
    "align": "justify",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "medium",
    "spacing": "0 0 1em 0",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 400,
  },
  "quote": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "medium",
    "spacing": "",
    "style": "italic",
    "tracking": "",
    "transform": "none",
    "weight": 400,
  },
  "strong": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "medium",
    "spacing": "",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 700,
  },
  "text": Object {
    "align": "left",
    "decoration": Object {
      "line": "none",
      "style": "",
    },
    "family": "Helvetica, sans-serif",
    "indent": "",
    "kerning": "normal",
    "letting": "1.6em",
    "size": "medium",
    "spacing": "",
    "style": "normal",
    "tracking": "",
    "transform": "none",
    "weight": 400,
  },
}
`;
